Do some using statements.

In [1]:
using static Microsoft.DotNet.Interactive.Formatting.PocketViewTags;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive;
//using XPlot.Plotly.Interactive;
using System.Collections.Generic;
using System.IO;
using System;
using System.Text.RegularExpressions;



See if the input text exists.

In [2]:
var fileExists = File.Exists("input.txt");
if (fileExists)
{
    Console.WriteLine("This file exists.  Nothing has blown up.....yet.");
}

This file exists.  Nothing has blown up.....yet.


Load the input into a string array

In [3]:
string[] input = System.IO.File.ReadAllLines("input.txt");
string[] sample = System.IO.File.ReadAllLines("sample.txt");
sample

[ Card 1: 41 48 83 86 17 | 83 86 6 31 17 9 48 53, Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19, Card 3: 1 21 53 59 44 | 69 82 63 72 16 21 14 1, Card 4: 41 92 73 84 69 | 59 84 76 51 58 5 54 83, Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36, Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11 ]

In [4]:
// get one string to test.
string sample0 = sample[0]; 



public class ScratchCard
{
    public int CardNumber;
    public List<int> WinningNumbers;
    public List<int> PotentialNumbers;
    public int Matches;
    public double Points;
    public List<int> BonusTicketNumbers;

    public ScratchCard(int cardNumber, List<int> winners, List<int> potentials)
    {
        this.CardNumber = cardNumber;
        this.WinningNumbers = winners;
        this.PotentialNumbers = potentials;
        this.Matches = winners.Intersect(potentials).Count();
        this.BonusTicketNumbers =  new List<int>();
        if(this.Matches == 0)
        {
            this.Points = 0;
        }
        else
        {
            this.Points = Math.Pow(2, this.Matches-1);
            int freeCards = this.Matches;
            for(int i = 1; i<=freeCards; i++)
            {
                this.BonusTicketNumbers.Add(this.CardNumber + i);
            }            
        }

    }
}

public List<int> GetNumbers(string str)
{
    List<int> numbers = new List<int>();
    Regex regex = new Regex(@"\b\d{1,2}\b", RegexOptions.Compiled);
    MatchCollection matches = regex.Matches(str);
    foreach (Match match in matches)
    {
        bool result = int.TryParse(match.Value, out int num);
        if (result)
        {
            numbers.Add(num);
        }
    }
    return numbers;
}

public int GetCardNumber(string cardString)
{
    string digits = Regex.Replace(cardString, "[^0-9]", "");
    int cardNum;
    bool success = int.TryParse(digits, out cardNum);
    //Console.WriteLine("CardNumber: " + cardNum);
    return cardNum;
}



In [5]:
// Part 1
List<ScratchCard> cards = new List<ScratchCard>();
double sum = 0;
foreach(string str in input)
{
    // Split stuff out.
    string[] firstSplit = str.Split(":");
    string CardInfo = firstSplit[0];
    string[] secondSplit = firstSplit[1].Split("|");
    string winningNumberString = secondSplit[0];
    string potentialNumberString = secondSplit[1];
    //Console.WriteLine("Potential Numbers: " +  potentialNumberString);
    //Convert strings to useful data types.
    List<int> winningNumbers = GetNumbers(winningNumberString);
    List<int> potentialNumbers = GetNumbers(potentialNumberString);
    int cardNumber = GetCardNumber(CardInfo);
    //Create the card
    ScratchCard card = new ScratchCard(cardNumber, winningNumbers, potentialNumbers);
    cards.Add(card);
    sum += card.Points;
}
//cards
sum

25571

Part 2: There's no such thing as "points". Instead, scratchcards only cause you to win more scratchcards equal to the number of winning numbers you have.

Specifically, you win copies of the scratchcards below the winning card equal to the number of matches. So, if card 10 were to have 5 matching numbers, you would win one copy each of cards 11, 12, 13, 14, and 15.

In [6]:
// Part 2
List<ScratchCard> cards = new List<ScratchCard>();
double sum = 0;
foreach(string str in input)
{
    // Split stuff out.
    string[] firstSplit = str.Split(":");
    string CardInfo = firstSplit[0];
    string[] secondSplit = firstSplit[1].Split("|");
    string winningNumberString = secondSplit[0];
    string potentialNumberString = secondSplit[1];
    //Console.WriteLine("Potential Numbers: " +  potentialNumberString);
    //Convert strings to useful data types.
    List<int> winningNumbers = GetNumbers(winningNumberString);
    List<int> potentialNumbers = GetNumbers(potentialNumberString);
    int cardNumber = GetCardNumber(CardInfo);
    //Create the card
    ScratchCard card = new ScratchCard(cardNumber, winningNumbers, potentialNumbers);
    cards.Add(card);
    sum += card.Points;
}
cards


index value 0 Submission#5+ScratchCard CardNumber 1 WinningNumbers [ 5, 37, 16, 3, 56, 11, 23, 72, 7, 8 ] PotentialNumbers [ 3, 79, 35, 45, 72, 69, 15, 14, 48, 88, 96, 37, 11, 75, 83, 56, 23, 7, 16, 50 ... (5 more) ] Matches 8 Points 128 BonusTicketNumbers [ 2, 3, 4, 5, 6, 7, 8, 9 ] 1 Submission#5+ScratchCard CardNumber 2 WinningNumbers [ 1, 45, 93, 96, 65, 88, 78, 15, 27, 26 ] PotentialNumbers [ 5, 84, 62, 63, 45, 61, 1, 80, 88, 77, 40, 51, 73, 21, 32, 98, 74, 59, 97, 9 ... (5 more) ] Matches 4 Points 8 BonusTicketNumbers [ 3, 4, 5, 6 ] 2 Submission#5+ScratchCard CardNumber 3 WinningNumbers [ 9, 99, 34, 44, 37, 16, 67, 43, 41, 83 ] PotentialNumbers [ 43, 41, 5, 69, 90, 50, 34, 94, 86, 59, 98, 16, 99, 28, 44, 37, 47, 57, 7, 14 ... (5 more) ] Matches 10 Points 512 BonusTicketNumbers [ 4, 5, 6, 7, 8, 9, 10, 11, 12, 13 ] 3 Submission#5+ScratchCard CardNumber 4 WinningNumbers [ 45, 99, 64, 82, 57, 9, 56, 17, 78, 7 ] PotentialNumbers [ 75, 56, 30, 88, 64, 1, 98, 27, 9, 57, 7, 6, 77, 44, 17, 78, 82, 99, 16, 91 ... (5 more) ] Matches 10 Points 512 BonusTicketNumbers [ 5, 6, 7, 8, 9, 10, 11, 12, 13, 14 ] 4 Submission#5+ScratchCard CardNumber 5 WinningNumbers [ 76, 80, 42, 88, 26, 56, 79, 63, 6, 37 ] PotentialNumbers [ 16, 4, 40, 34, 46, 76, 67, 69, 1, 54, 5, 55, 59, 24, 78, 29, 26, 9, 51, 44 ... (5 more) ] Matches 4 Points 8 BonusTicketNumbers [ 6, 7, 8, 9 ] 5 Submission#5+ScratchCard CardNumber 6 WinningNumbers [ 59, 23, 88, 38, 49, 16, 24, 18, 22, 89 ] PotentialNumbers [ 52, 25, 88, 27, 23, 79, 22, 84, 72, 80, 39, 17, 49, 96, 56, 60, 44, 45, 16, 63 ... (5 more) ] Matches 6 Points 32 BonusTicketNumbers [ 7, 8, 9, 10, 11, 12 ] 6 Submission#5+ScratchCard CardNumber 7 WinningNumbers [ 81, 1, 37, 6, 20, 76, 3, 31, 93, 83 ] PotentialNumbers [ 74, 32, 25, 76, 43, 87, 52, 93, 47, 85, 83, 31, 17, 72, 6, 99, 1, 36, 20, 81 ... (5 more) ] Matches 10 Points 512 BonusTicketNumbers [ 8, 9, 10, 11, 12, 13, 14, 15, 16, 17 ] 7 Submission#5+ScratchCard CardNumber 8 WinningNumbers [ 74, 73, 65, 29, 66, 47, 43, 11, 24, 38 ] PotentialNumbers [ 5, 3, 1, 88, 29, 11, 49, 67, 47, 33, 31, 61, 63, 75, 84, 35, 18, 71, 66, 92 ... (5 more) ] Matches 4 Points 8 BonusTicketNumbers [ 9, 10, 11, 12 ] 8 Submission#5+ScratchCard CardNumber 9 WinningNumbers [ 67, 68, 8, 74, 17, 11, 28, 47, 96, 2 ] PotentialNumbers [ 85, 7, 37, 33, 15, 18, 91, 96, 4, 67, 16, 47, 28, 26, 80, 52, 17, 97, 68, 8 ... (5 more) ] Matches 10 Points 512 BonusTicketNumbers [ 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 ] 9 Submission#5+ScratchCard CardNumber 10 WinningNumbers [ 91, 22, 85, 35, 47, 26, 99, 39, 72, 38 ] PotentialNumbers [ 5, 7, 12, 14, 62, 93, 61, 56, 82, 4, 1, 51, 86, 36, 43, 29, 50, 75, 68, 25 ... (5 more) ] Matches 0 Points 0 BonusTicketNumbers [ ] 10 Submission#5+ScratchCard CardNumber 11 WinningNumbers [ 4, 53, 44, 83, 23, 84, 40, 55, 69, 82 ] PotentialNumbers [ 24, 48, 11, 37, 60, 76, 41, 29, 58, 39, 45, 88, 95, 67, 49, 28, 36, 35, 86, 33 ... (5 more) ] Matches 0 Points 0 BonusTicketNumbers [ ] 11 Submission#5+ScratchCard CardNumber 12 WinningNumbers [ 40, 45, 87, 58, 72, 59, 89, 55, 20, 91 ] PotentialNumbers [ 7, 62, 2, 91, 59, 78, 4, 44, 25, 24, 57, 94, 79, 75, 51, 54, 55, 90, 83, 30 ... (5 more) ] Matches 3 Points 4 BonusTicketNumbers [ 13, 14, 15 ] 12 Submission#5+ScratchCard CardNumber 13 WinningNumbers [ 92, 58, 35, 96, 84, 62, 31, 65, 95, 5 ] PotentialNumbers [ 22, 52, 84, 98, 62, 31, 75, 7, 12, 78, 51, 91, 37, 58, 46, 85, 21, 61, 95, 49 ... (5 more) ] Matches 7 Points 64 BonusTicketNumbers [ 14, 15, 16, 17, 18, 19, 20 ] 13 Submission#5+ScratchCard CardNumber 14 WinningNumbers [ 75, 37, 41, 53, 12, 77, 97, 6, 54, 29 ] PotentialNumbers [ 52, 65, 46, 94, 20, 6, 76, 75, 70, 83, 29, 93, 64, 1, 12, 58, 89, 49, 26, 16 ... (5 more) ] Matches 5 Points 16 BonusTicketNumbers [ 15, 16, 17, 18, 19 ] 14 Submission#5+ScratchCard CardNumber 15 WinningNumbers [ 20, 88, 22, 99, 28, 87, 16, 78, 70, 71 ] PotentialNumbers [ 20, 89, 56, 27, 61, 32, 53, 22, 78, 3, 54, 28, 70, 64, 33, 24, 23, 17, 5

In [7]:
public class ScratchCardSkeleton
{
    public int CardNumber;
    public int[] BonusTicketNumbers;

    public ScratchCardSkeleton(int cardNumber, List<int> bonuses)
    {
        this.CardNumber = cardNumber;
        this.BonusTicketNumbers =  bonuses.ToArray<int>();
    }

    public ScratchCardSkeleton(int cardNumber, int[] bonuses)
    {
        this.CardNumber = cardNumber;
        this.BonusTicketNumbers =  bonuses;
    }
}

//convert cards to skeleton version of cards.
List<ScratchCardSkeleton> scs = new List<ScratchCardSkeleton>();
Dictionary<int, int[]> cardDict =  new Dictionary<int, int[]>(); 
foreach(ScratchCard sc in cards)
{
    ScratchCardSkeleton ss =  new ScratchCardSkeleton(sc.CardNumber, sc.BonusTicketNumbers);
    cardDict[ss.CardNumber] = ss.BonusTicketNumbers;
    scs.Add(ss);
}

Console.WriteLine("Skeleton created");

int ScratchCardCount = 0;
while(scs.Count > 0)
{
    ScratchCardCount++;
    foreach(int cardNum in scs[0].BonusTicketNumbers)
    {
        ScratchCardSkeleton newFreeCard = new ScratchCardSkeleton(cardNum, cardDict[cardNum]);
        scs.Add(newFreeCard);
    }
    //Console.WriteLine("Card Process count: " + cards.Count);
    scs.Remove(scs[0]);
    //Console.WriteLine("Card Process count: " + cards.Count);
    //scs.OrderBy(o=>o.CardNumber).ToList();
    if (ScratchCardCount % 1000 == 0)
    {
        Console.WriteLine($"Card Count: {ScratchCardCount} - Remaining cards: {scs.Count}");
    }

}


//sum
ScratchCardCount

Skeleton created
Card Count: 1000 - Remaining cards: 3702
Card Count: 2000 - Remaining cards: 6981
Card Count: 3000 - Remaining cards: 9730
Card Count: 4000 - Remaining cards: 12160
Card Count: 5000 - Remaining cards: 15190
Card Count: 6000 - Remaining cards: 17629
Card Count: 7000 - Remaining cards: 19573
Card Count: 8000 - Remaining cards: 22811
Card Count: 9000 - Remaining cards: 25615
Card Count: 10000 - Remaining cards: 27202
Card Count: 11000 - Remaining cards: 29361
Card Count: 12000 - Remaining cards: 32186
Card Count: 13000 - Remaining cards: 33462
Card Count: 14000 - Remaining cards: 35205
Card Count: 15000 - Remaining cards: 37326
Card Count: 16000 - Remaining cards: 40143
Card Count: 17000 - Remaining cards: 42302
Card Count: 18000 - Remaining cards: 45317
Card Count: 19000 - Remaining cards: 47435
Card Count: 20000 - Remaining cards: 48768
Card Count: 21000 - Remaining cards: 49901
Card Count: 22000 - Remaining cards: 51905
Card Count: 23000 - Remaining cards: 54035
Card C

8805731

8805731 is right! Avoiding LINQ and complicated object sorting made it fast enough to run everything.